numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -1925,6 +1925,7 @@ typedef void (PyDataMem_EventHookFunc)(void *inp, void *outp, size_t size,
          * (structured and object are exceptions).
          * This should potentially become a weak mapping in the future.
          */
+        PyObject *within_dtype_castingimpl;
         PyObject *castingimpls;
```

Comment:
The comment remains unchanged. Since these are C-level objects, would we gain anything by making them a pointer-to-function rather than a python `PyObject*` ? If they need to be exposed to python, they could be wrapped before being returned. For user-defined dtypes, the function could use the python C-API to call back out to python,

Reply:
These are `ArrayMethod`s.  Somehow, I don't have a clear mental model for making something that may be defined in Python a bare C-struct (or the advantage over it just beeing it a `PyObject *`). If we may want to be able to delete a dtype (and with it these casts), it seems to me that `PyObject *` is more natural than a plain struct.

I somewhat understand the desire to keep out Python where possible, but not sure here. Maybe I am just missing the right design patterns for this?

Reply:
I guess we can revisit it in the future.